File: Elixir/problem_5.ex
# Problem 5: smallest multiple
# https://projecteuler.net/problem=6

defmodule Divisibility do
  def smallest_multiple(x) do
    recursive_lcm(x, x-1)
      |> round
  end

  def recursive_lcm(x, 1), do: x
  def recursive_lcm(x, y) do
    recursive_lcm(lcm(x, y), y - 1)
  end

  def lcm(a, b) do
    a*b / gcd(a,b)
      |> round
  end

  def gcd(x, 0), do: x
  def gcd(x, y), do: gcd(y, rem(x, y))
end

# Result:
# > Divisibility.smallest_multiple(20)
# 232792560

File: R/problem_5.R
# Problem 5: smallest multiple
# https://projecteuler.net/problem=5

smallest_multiple <- function(n){
    result <- 1
    for(i in 2:n){
        result <- lcm(result, i)
    }
    result
}
lcm <- function(x, y){
    x * y / gcd(x, y)
}
gcd <- function(x, y){
  r <- x%%y
  return(ifelse(r, gcd(y, x%%y), y))
}

# Result:
# > smallest_multiple(20)
